pkg/cloudprovider/aws/amifamily/ami.go

```diff
@@ -19,30 +19,170 @@ import (
 	"fmt"
 
 	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/service/ec2/ec2iface"
 	"github.com/aws/aws-sdk-go/service/ssm"
 	"github.com/aws/aws-sdk-go/service/ssm/ssmiface"
+	"github.com/mitchellh/hashstructure/v2"
 	"github.com/patrickmn/go-cache"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
 	"knative.dev/pkg/logging"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 
+	"github.com/samber/lo"
+
+	"github.com/aws/karpenter/pkg/apis/awsnodetemplate/v1alpha1"
+	"github.com/aws/karpenter/pkg/apis/provisioning/v1alpha5"
 	"github.com/aws/karpenter/pkg/cloudprovider"
+	awsv1alpha1 "github.com/aws/karpenter/pkg/cloudprovider/aws/apis/v1alpha1"
+	"github.com/aws/karpenter/pkg/scheduling"
+	"github.com/aws/karpenter/pkg/utils/functional"
+	"github.com/aws/karpenter/pkg/utils/sets"
 )
 
 type AMIProvider struct {
-	cache *cache.Cache
-	ssm   ssmiface.SSMAPI
+	ssmCache   *cache.Cache
+	ec2Cache   *cache.Cache
+	ssm        ssmiface.SSMAPI
+	kubeClient client.Client
+	ec2api     ec2iface.EC2API
 }
 
 // Get returns a set of AMIIDs and corresponding instance types. AMI may vary due to architecture, accelerator, etc
-func (p *AMIProvider) Get(ctx context.Context, instanceType cloudprovider.InstanceType, ssmQuery string) (string, error) {
-	if id, ok := p.cache.Get(ssmQuery); ok {
+// If AMI overrides are specified in the AWSNodeTemplate, then only those AMIs will be chosen.
+func (p *AMIProvider) Get(ctx context.Context, provider *awsv1alpha1.AWS, nodeRequest *cloudprovider.NodeRequest, options *Options, amiFamily AMIFamily) (map[string][]cloudprovider.InstanceType, error) {
+	amiIDs := map[string][]cloudprovider.InstanceType{}
+	amiRequirements, err := p.getAMIRequirements(ctx, nodeRequest.Template.ProviderRef)
+	if err != nil {
+		return nil, err
+	}
+	var amiID string
```

Comment:
This stray variable is a bit odd to me, since you shadow it on line 63, but don't shadow it on line 64. I'm surprised the linter isn't complaining, since we only allow shadowing ctx/err.